### PR TITLE
Simplify bulk request execution 

### PR DESCRIPTION
**Pull request by areek**
Currently, bulk item requests can be any ActionRequest, this PR
restricts bulk item requests to DocumentRequest. This simplifies
handling failures during bulk requests. Additionally, a new enum
is added to DocumentRequest to represent the intended operation
to be performed by a document request (`create`, `index`, `update` 
and `delete`), which was previously represented with a mix of strings 
and index request operation type.
Now, index request operation type reuses the new enum to specify 
whether the request should `create` or `index` a document.
Restricting bulk requests to DocumentRequest further simplifies
execution of shard-level bulk operations to use the same failure 
handling for index, delete and update operations.
This PR also fixes a bug which executed delete operations twice for 
replica copies while executing bulk requests.

relates https://github.com/elastic/elasticsearch/pull/19105

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java [hunk 1]
```diff
@@ -322,7 +328,7 @@ public void readFrom(StreamInput in) throws IOException {
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         out.writeVInt(id);
-        out.writeString(opType);
+        out.writeByte(opType.getId());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java]: we need bwc here too

**Review by bleskes — CHANGES_REQUESTED** [review]
I went through this _very_ carefully and it looks great. It's an amazing restructuring and I'm happy you took it on yourself. I left some minor comments. I also think we can potentially simplify further and make the DocumentRequest inherit from WriteReplicationRequest  (and call DocWriteRequest) . I'm not sure but it would be great if you can give it a go.

File: core/src/main/java/org/elasticsearch/action/DocumentRequest.java [hunk 2]
```diff
@@ -70,4 +72,78 @@
...
+        UPDATE(2),
+        /** Deletes a document */
+        DELETE(3);
+
+        private final byte op;
+        private final String lowercase;
+
+        OpType(int op) {
+            this.op = (byte) op;
+            this.lowercase = this.toString().toLowerCase(Locale.ENGLISH);
```
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/DocumentRequest.java]: changed it to ROOT, but there are a few places (e.g. `DocWriteResponse`) where we use ENGLISH too. maybe we should change it to ROOT as well?

File: core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java [hunk 3]
```diff
@@ -577,15 +574,17 @@ public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
         waitForActiveShards.writeTo(out);
         out.writeVInt(requests.size());
-        for (ActionRequest<?> request : requests) {
+        for (DocumentRequest<?> request : requests) {
             if (request instanceof IndexRequest) {
```
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java]: done

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 4]
```diff
@@ -132,203 +126,84 @@ protected boolean resolveIndex() {
...
+                                                     Translog.Location location, int requestIndex) {
+        preVersions[requestIndex] = request.items()[requestIndex].request().version();
+        preVersionTypes[requestIndex] = request.items()[requestIndex].request().versionType();
+        DocumentRequest.OpType opType = request.items()[requestIndex].request().opType();
         try {
-            WriteResult<IndexResponse> result = shardIndexOperation(request, indexRequest, metaData, indexShard, true);
-            location = locationToSync(location, result.getLocation());
+            WriteResult<? extends DocWriteResponse> writeResult = innerExecuteBulkItemRequest(metaData, indexShard,
+                request, requestIndex);
+            if (writeResult.getResponse().getResult() != DocWriteResponse.Result.NOOP) {
```
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: I wanted to be more strict as only a noop result can be valid with a null location. changed it to be an assert instead.

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 5]
```diff
@@ -132,203 +126,84 @@ protected boolean resolveIndex() {
...
             }
-            logFailure(e, "index", request.shardId(), indexRequest);
-            // if its a conflict failure, and we already executed the request on a primary (and we execute it
-            // again, due to primary relocation and only processing up to N bulk items when the shard gets closed)
-            // then just use the response we got from the successful execution
-            if (item.getPrimaryResponse() != null && isConflictException(e)) {
-                setResponse(item, item.getPrimaryResponse());
+            BulkItemRequest item = request.items()[requestIndex];
+            DocumentRequest<?> documentRequest = item.request();
+            if (ExceptionsHelper.status(e) == RestStatus.CONFLICT) {
```
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: fixed

File: core/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java (hunk 1 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java]: bwc added

**Comment by areek** [comment]
Thanks @bleskes for the review :). I addressed all the minor comments. 

> I also think we can potentially simplify further and make the DocumentRequest inherit from WriteReplicationRequest (and call DocWriteRequest)

I really like the idea of making DocumentRequest inherit from WriteReplicationRequest. While giving this a go, I refactored update operation to be a replication operation (updates are a DocumentWriteRequest but does't use the replication operation instead delegates to index and delete operations for replication). This cleans up the dedicated transport update action (`TransportInstanceSingleOperationAction`) in favour of reusing replication action and duplicate code for update retry logic and translation. The one big downside is it would be difficult to implement wire bwc with 5.0. WDYT?

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 6]
```diff
@@ -338,105 +214,56 @@ private void setResponse(BulkItemRequest request, BulkItemResponse response) {
...
+                            (updateRequest.fields() != null && updateRequest.fields().length > 0)) {
+                        Tuple<XContentType, Map<String, Object>> sourceAndContent = XContentHelper.convertToMap(indexSourceAsBytes, true);
+                        update.setGetResult(updateHelper.extractGetResult(updateRequest, updateRequest.concreteIndex(), indexResponse.getVersion(), sourceAndContent.v2(), sourceAndContent.v1(), indexSourceAsBytes));
+                    } else {
+                        update.setGetResult(null);
                     }
-                    return new UpdateResult(translate, indexRequest, retry, cause, null);
+                } else {
+                    assert translate.getResponseResult() == DocWriteResponse.Result.UPDATED;
+                    update.setGetResult(updateHelper.extractGetResult(updateRequest, updateRequest.concreteIndex(), indexResponse.getVersion(), translate.updatedSourceAsMap(), translate.updateSourceContentType(), indexSourceAsBytes));
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: why do we now return always the get result on updates? it seems to be different [before](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java#L251)

**Review by bleskes — COMMENTED** [review]
Thanks @areek . Let's leave the merger/inheritance of DocumentRequest and ReplicationRequest alone. None of the solutions seem ideal (update request is not a replication request and the fact it's running on the primary is not a good thing). This PR is great enough and it would be a shame to delay it. We can revisit / evaluate again when have a better idea.

I do have two small asks (next to a question I left in the comments)
- Can we rename `DocumentRequest` to `DocWriteRequest`? it will be inline with the `DocWriteResponse` we have today.
- It seems we can get rid of the `<?>` in all the `DcoumentRequest<?>` - I think it will make it simpler to read. See https://gist.github.com/bleskes/7ec26991d7be6c816103b8cac65353a4

**Comment by bleskes** [comment]
It also seems CI isn't happy.

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java (hunk 6 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: I changed it to match the logic in [TransportUpdateAction.shardOperation](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/update/TransportUpdateAction.java#L175) but that caused CI failures, I reverted the change to be the same as update operation in bulk

**Comment by areek** [comment]
Thanks @bleskes for the feedback. I updated the PR, addressing your comments and the CI is happy. Could you take a look

**Review by bleskes — APPROVED** [review]

**Comment by bleskes** [comment]
LGTM. Awesome @areek

**Pull request merged**